File: solved/median-of-two-sorted-arrays.c
double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size)
{
    int m = (nums1Size + nums2Size) / 2;
    int i = 0;
    int j = 0;
    int k = 0;
    double r;
    int *a = malloc((m + 1) * sizeof(int));

    while (i < nums1Size && j < nums2Size && k <= m) {
        if (nums1[i] < nums2[j]) {
            a[k++] = nums1[i++];
        } else {
            a[k++] = nums2[j++];
        }
    }
    while (i < nums1Size && k <= m) {
        a[k++] = nums1[i++];
    }
    while (j < nums2Size && k <= m) {
        a[k++] = nums2[j++];
    }
    r = ((nums1Size + nums2Size) & 1) ? a[m] : (a[m] + a[m - 1]) / 2.0;
    free(a);
    return r;
}

File: solved/median-of-two-sorted-arrays.rs
impl Solution {
    fn find_kth(nums1: &[i32], nums2: &[i32], k: usize) -> i32 {
        match (nums1.len(), nums2.len()) {
            (0, 0) => 0,
            (0, _) => nums2[k],
            (_, 0) => nums1[k],
            (n, m) => if k == 0 {
                nums1[0].min(nums2[0])
            } else {
                let half_k = (k - 1) / 2;
                let i = if half_k >= n { n - 1 } else { half_k };
                let j = if half_k >= m { m - 1 } else { half_k };
                if nums1[i] < nums2[j] {
                    Self::find_kth(&nums1[i + 1..], nums2, k - i - 1)
                } else {
                    Self::find_kth(nums1, &nums2[j + 1..], k - j - 1)
                }
            }
        }
    }

    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let k = (nums1.len() + nums2.len()) / 2;
        let v = Self::find_kth(&nums1, &nums2, k);
        if (nums1.len() + nums2.len()) % 2 == 0 && k > 0 {
            let v1 = Self::find_kth(&nums1, &nums2, k - 1);
            (v + v1) as f64 / 2f64
        } else {
            v as f64
        }
    }
}
